feat(#62): markdown support for labels, hints, and constraint messages

garethbowen opened the pull request:
Closes #62 

### I have verified this PR works in these browsers (latest versions):

- [ ] Chrome
- [x] Firefox
- [ ] Safari (macOS)
- [ ] Safari (iOS)
- [ ] Chrome for Android
- [ ] Not applicable

### What else has been done to verify that this works as intended?

Automated tests

### Why is this the best possible solution? Were any other approaches considered?

### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Markdown processing runs automatically on all labels, so forms that include markdown styling accidentally will look different, for example, questions starting with a number will now be rendered as an ordered list which means they will be indented when they weren't before.

### Do we need any specific form for testing your changes? If so, please attach one.

From this branch: `/packages/common/src/fixtures/notes/3-notes-with-markdown.xml`

### What's changed

garethbowen commented on packages/common/src/fixtures/select/1-static-selects.xml:
```diff
@@ -13,10 +13,10 @@
         <translation lang="French">
           <text id="fruit:label">
-            <value>1. Sélectionnez un fruit</value>
+            <value>`1. Sélectionnez un fruit`</value>
```
These had to be changed because without the backtick they were rendering as an ordered list which broke a test.

garethbowen commented on packages/web-forms/src/components/form-elements/ControlLabel.vue:
```diff
@@ -1,12 +1,12 @@
-	<label :for="question.nodeId">
-		<span v-if="question.currentState.required" class="required">* </span>
-		<span>{{ props.question.currentState.label?.asString }}</span>
+	<label :for="question.nodeId" :class="{ required: question.currentState.required }">
```
I've changed the `*` to be in the `:before` content because that's more semantic (the `*` is a stylistic representation of the required state, not part of the label text). It was also needed to align the `*` with rich content (eg: `<h1>`) which had some top padding.

garethbowen commented:
@latin-panda Review please! It's a fairly big change and much more complex than I'd like - feedback welcome!

garethbowen commented:
My todo list...

- [ ] make sure the styling works when used in central
- [ ] go through the issue again and make sure I haven't missed anything
- [ ] get this through qa

garethbowen commented on packages/web-forms/src/components/form-elements/ControlHint.vue:
```diff
@@ -1,11 +1,12 @@
 <template>
 	<div v-if="question.currentState.hint" class="hint">
-		{{ question.currentState.hint?.asString }}
+		<MarkdownBlock v-for="(elem, index) in question.currentState.hint?.formatted" :key="index" :elem="elem" />
```
I propose we leave this as is - the Central team have decided that anything wider than 77ch is not readable. In this case it doesn't actually break web-forms and if they want to change it it can be reset on their side. If we start including overrides for everything any customer may have set in the parent context we'll end up with the world's biggest reset file (eg: we'd also have to do something about the following line `p:lang(ja), p:lang(zh) { max-width: 54ch; }`). What do you think?

latin-panda replied:
Okay, in WF standalone, the text is fully expanded. This only affects WF in Central.

garethbowen commented on packages/web-forms/src/assets/styles/reset.scss:
```diff
@@ -34,16 +34,27 @@ strong {
 span,
 label {
-	font-weight: normal;
+	font-weight: inherit;
```
Yes :sob: 

So the reset wasn't really working - even for the question labels we had specific weight override in ControlLabel.vue. I've added the same in `select-options.scss`.

garethbowen commented on packages/web-forms/src/assets/styles/reset.scss:
```diff
@@ -34,16 +34,27 @@ strong {
 
+label,
+.hint {
+	p {
```
Oof. I'm not sure why this changed - I would expect it to be broken in `main` too. Bottom line is Central defines a `.label { color: white }` and we don't override it either in `main` or this branch. Fixed by overriding.

latin-panda replied:
Hmm, I wonder if it's some new code on their affecting Web Forms.

garethbowen commented on packages/xforms-engine/src/client/MarkdownNode.ts:
```diff
@@ -0,0 +1,52 @@
+export interface StyleProperty {
+	readonly color: string | undefined;
+	readonly 'font-family': string | undefined;
+	readonly 'text-align': 'center' | 'left' | 'right' | undefined;
```
This is caused by the `max-width: 77ch` as above. The reason it doesn't line up with the label is the font-size in the label is bigger and therefore the `max-width` is wider. As above, I propose leaving this for Central to define whatever max-width they want.

garethbowen commented on packages/web-forms/src/components/common/MarkdownBlock.vue:
```diff
@@ -0,0 +1,55 @@
+	<span v-else-if="elem.role === 'html'" v-html="purify(elem)" />
+
+	<!-- link -->
+	<a v-else-if="elem.role === 'parent' && elem.elementName === 'a'" :href="getUrl(elem)" target="_blank">
```
Yes precisely. I wanted to avoid actually parsing the HTML and just sanitizing it which means we're limited about what changes we can make to it. In this case, the person building the form would have to specify the `target` attribute. Is that ok?

garethbowen commented on packages/web-forms/src/components/form-elements/ControlLabel.vue:
```diff
@@ -1,12 +1,30 @@
-		<span>{{ props.question.currentState.label?.asString }}</span>
+	<label :for="question.nodeId" :class="{ required: question.currentState.required }">
+		<MarkdownBlock v-for="(elem, index) in text" :key="index" :elem="elem" />
+		<div v-if="image || video || audio" class="media-content">
```
I'd like to put this out of scope, because it may require design input separate testing, etc. Should I remove the image handling code that I added pending dev on the separate issue, or do you think I leave it in there because it's better than nothing?

garethbowen replied:
Will do.

garethbowen replied:
Issue raised: #525

latin-panda commented:
@srujner Collect displays selections as a comma-separated list, which better conveys what was selected rather than how many items were chosen.

<img width="300" src="https://github.com/user-attachments/assets/4cea4ed5-db14-4b4b-a132-fc8c128bb0d7" />

garethbowen commented:
> Text should be centered and is not 

Fixed by applying a width to the option label so text alignment works as expected.

> One of the fonts looks Ok on Iphone device, but in browser it's broken 

Fonts in browsers are difficult because they rely on the underlying OS to provide the font. Ultimately this comes down to licensing because most fonts aren't open so Windows doesn't ship with Helvetica (for example). In this case the label is `font-family: fantasy` which is clearly being rendered very differently. Everything is working perfectly as far as the markdown is concerned, it's just the font referred to by `fantasy` is quite different on different OSes.

> Looks like Underline does not work in webforms 

You're absolutely right. Fixed now, along with `<i>`.

> If the answer is a long text, then you can't really see what is selected, you have to expand the select.

I've left this as out of scope for now.

@srujner I've fixed the centered options, and the underline, and updated the netlify project with the latest version. Would you mind giving it another look?
